torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2247,88 +2385,220 @@ def _load_sharded_state_dict(
+        input: Any,
+    ):
+        """
+        Runs the pre-forward logic. This includes an opportunity to reshard
+        currently unsharded parameters such as those from previous forwards,
+        unsharding the parameters for the current forward, and registering
+        post-backward hooks on these current parameters.
+
+        For the non-recursive wrapping path, the original parameters are
+        unflattened here in the pre-forward, while in the recursive wrapping
+        path, the original parameters are unflattened in the beginning of the
+        :class:`FlattenParamsWrapper` 's ``forward()``.
+
+        Args:
+            handles (List[FlatParamHandle]): Handles giving the parameters
+                used in the current forward.
+            reshard_fn (Optional[Callable]): A callable to reshard any
+                currently unsharded parameters e.g. from previous forwards or
+                ``None`` to not do any resharding.
+            module (nn.Module): Unused; expected by the hook signature.
+            input (Any): Unused; exepcted by the hook signature.
+        """
         with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
             self._lazy_init()
             self._wait_for_previous_optim_step()
-
-            # Start of a forward pass.
             self.training_state = TrainingState_.FORWARD
-            if self._is_root:
-                # TODO: disabling side stream for tensor copies for now, investigate
-                # perf with it on / off.
-                # Place inputs on compute_device. This is a noop if inputs are already
-                # on compute_device. Note that when device_id is specified,
-                # device_id == self.compute_device is guaranteed.
-                # TODO: for mixed precision, move inputs to right device + cast might
-                # be done in one go for performance.
-                args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
-                args = args[0]
-                kwargs = kwargs[0]
-
-            # Cast inputs to their mixed precision type.
-            if (
-                self._is_root
-                and self._mixed_precision_enabled_for_params()
-            ):
-                input_dtype = self.mixed_precision.param_dtype
-                args, kwargs = self._cast_fp_inputs_to_precision(
-                    input_dtype, *args, **kwargs
-                )
-
-            # Only rebuilding full params when the params are not prefetched in previous layers
-            if not self._forward_full_params_prefetched:
-                self._rebuild_full_params()
-            self._forward_full_params_prefetched = False
-            # Wait for all_gather full parameters to finish before computation
+            if reshard_fn is not None:
+                reshard_fn()
+            params = [handle.flat_param for handle in self._handles]
+            self._rebuild_full_params(params)
```

Comment:
for non-recursive policy, will this rebuild parameters of all modules?

Reply:
I do not think so. This should only rebuild the `FlatParameters` that contain parts needed for the module's forward.

See `_register_pre_forward_hooks()` below:
```
reshard_fn = functools.partial(self._pre_forward_reshard, module_flat_param_handles)
```

`module_flat_param_handes` are exactly the handles needed for the (sub)module's forward. This should typically not be all of the handles.

Reply:
well, I think in line 2468 it should be 
```python
            params = [handle.flat_param for handle in handles]
```
`self. _handles ` contains all the handles I think, and currently you are rebuilding all params.

Reply:
Good catch :)